minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,635 @@
+
+
+from __future__ import annotations
+
+from collections import namedtuple
+from itertools import product
+from typing import Callable, Generator, Iterable
+
+from minorminer.utils.zephyr.coordinate_systems import (
+    CartesianCoord,
+    ZephyrCoord,
+    cartesian_to_zephyr,
+    zephyr_to_cartesian,
+)
+from minorminer.utils.zephyr.plane_shift import PlaneShift
+
+ZShape = namedtuple("ZShape", ["m", "t"], defaults=(None, None))
+
+
+class Edge:
+    """Initializes an Edge with nodes x, y.
+
+    Args:
+        x : One endpoint of edge.
+        y : Another endpoint of edge.
+    """
+
+    def __init__(
+        self,
+        x,
+        y,
+    ) -> None:
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x, y):
+        if x < y:
+            return (x, y)
+        else:
+            return (y, x)
+
+    def __hash__(self):
+        return hash(self._edge)
+
+    def __getitem__(self, index: int) -> int:
+        return self._edge[index]
+
+    def __eq__(self, other: Edge):
+        return self._edge == other._edge
+
+    def __str__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+
+class ZEdge(Edge):
+    """Initializes a ZEdge with 'ZNode' nodes x, y
+
+    Args:
+        x (ZNode): One endpoint of edge.
+        y (ZNode): Another endpoint of edge.
+        check_edge_valid (bool, optional): Flag to whether check the validity of values and types of x, y.
+        Defaults to True.
+
+    Raises:
+        TypeError: If either of x or y is not 'ZNode'.
+        ValueError: If x, y do not have the same shape.
+        ValueError: If x, y are not neighbours in a perfect yield (quotient)
+        Zephyr graph.
+
+    Example 1:
+    >>> from zephyr_utils.node_edge import ZNode, ZEdge
+    >>> e = ZEdge(ZNode((3, 2)), ZNode((7, 2)))
+    >>> print(e)
+    ZEdge(ZNode(CartesianCoord(x=3, y=2, k=None)), ZNode(CartesianCoord(x=7, y=2, k=None)))
+    Example 2:
+    >>> from zephyr_utils.node_edge import ZNode, ZEdge
+    >>> ZEdge(ZNode((2, 3)), ZNode((6, 3))) # raises error, since the two are not neighbors
+    """
+
+    def __init__(
+        self,
+        x: ZNode,
+        y: ZNode,
+        check_edge_valid: bool = True,
+    ) -> None:
+        if check_edge_valid:
+            if not isinstance(x, ZNode) or not isinstance(y, ZNode):
+                raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")
+            if x.shape != y.shape:
+                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")
+            kind_found = False
+            for kind in ("internal", "external", "odd"):
+                if x.is_neighbor(y, nbr_kind=kind):
+                    kind_found = True
+                    break
+            if not kind_found:
+                raise ValueError(f"Expected x, y to be neighbours, got {x, y}")
```

Comment:
Could be simplified to:
```suggestion
            for kind in ("internal", "external", "odd"):
                if x.is_neighbor(y, nbr_kind=kind):
                    break
            else:
                raise ValueError(f"Expected x, y to be neighbours, got {x, y}")
```

Reply:
[9f7e5ab](https://github.com/dwavesystems/minorminer/pull/261/commits/9f7e5ab841c21b74bb7ddb875472aa38faf93a14)